AirLib/include/vehicles/multirotor/firmwares/simple_flight/firmware/AccelerationController.hpp

```diff
@@ -0,0 +1,128 @@
+#include "Params.hpp"
+#include "PidController.hpp"
+#include "common/common_utils/Utils.hpp"
+#include "interfaces/CommonStructs.hpp"
+#include "interfaces/IAxisController.hpp"
+#include "interfaces/IBoardClock.hpp"
+#include "interfaces/IGoal.hpp"
+#include "interfaces/IUpdatable.hpp"
+
+namespace simple_flight {
+
+class AccelerationController : public IAxisController,
+                               public IGoal // for internal child controller
+{
+public:
+  AccelerationController(Params *params, const IBoardClock *clock = nullptr)
+      : params_(params), clock_(clock) {}
+
+  virtual void initialize(unsigned int axis, const IGoal *goal,
+                          const IStateEstimator *state_estimator) override {
+    axis_ = axis;
+    goal_ = goal;
+    state_estimator_ = state_estimator;
+
+    child_mode_ = GoalMode::getUnknown();
+    // The output of this controller is an angle for roll and pitch and a
+    // throttle value. Throttle can directly be set as an output, but for the
+    // angles we need to cascade down to angle controllers.
+    switch (axis_) {
+    case 0:
+    case 1:
+      child_controller_ = std::make_unique<AngleLevelController>(params_, clock_);
+      child_mode_[axis_] = GoalModeType::AngleLevel;
+      break;
+    case 3:
+      // we output throttle directly, technically don't need this one
+      child_controller_ = std::make_unique<PassthroughController>();
+      child_mode_[axis_] = GoalModeType::Passthrough;
+      break;
+    default:
+      throw std::invalid_argument("Axis must be 0, 1 or 3. AccelerationController"
+                                  "controller cannot control yaw.");
+    }
+
+    // initialize child controller
+    child_controller_->initialize(axis_, this, state_estimator_);
+  }
+
+  virtual void reset() override {
+    IAxisController::reset();
+
+    child_controller_->reset();
+    child_goal_ = Axis4r();
+    output_ = TReal();
+  }
+
+  virtual void update() override {
+    IAxisController::update();
+
+    // Convert acceleration to linearised body frame
+    const Axis3r &goal_acc_world =
+        Axis4r::axis4ToXyz(goal_->getGoalValue(), true);
+    TReal yaw = state_estimator_->getAngles().yaw();
+    TReal ax = goal_acc_world.x() * cos(yaw) + goal_acc_world.y() * sin(yaw);
+    TReal ay = -goal_acc_world.x() * sin(yaw) + goal_acc_world.y() * cos(yaw);
+    TReal az = goal_acc_world.z();
+
+    // Convert acceleration setpoint to thrust vector
+    static constexpr float G = 9.81f;
+    static constexpr float MAX_ACC_XY = G / 1.41f;
+    // We allow maximum roll or pitch angles of ~45 deg.
+    ax = std::min(std::max(ax, -MAX_ACC_XY), MAX_ACC_XY);
+    ay = std::min(std::max(ay, -MAX_ACC_XY), MAX_ACC_XY);
+    Vector3r body_z = Vector3r(ax, ay, G).normalized();
+    // todo: add estimator instead of constant value
+    static constexpr TReal hover_thrust = 0.58f;
+    TReal collective_thrust = az * (hover_thrust / G) - hover_thrust;
```

Comment:
I don't think PX4 have got this quite right, as it assumes that the thrust creates 1g.  If az is non-zero (e.g. climbing) then you'll also get more ax & ay as a result.  Don't hold up this pull, but if you could add a "To do" to check later that would be good.
At a quick look I think the fix might be to add az to the calculation of body_z above (i.e. Vector3r(ax, ay, G - az).normalized(), but do correct if you disagree.

Reply:
You are right, I followed the math through and correcting the initial body_z vector as suggested will give more accurate acceleration in x. I'll add that since I am mentally already there that shouldn't take too much time.